File: TouchControl/src/Slider.java
import org.opencv.core.*;
import org.opencv.imgproc.Imgproc;

public abstract class Slider extends Touchable
{
	protected int numOfDivisions;
	protected int division;
	private boolean percentageVisible;
	private double divisionSize;

	private final static int DEFAULT_DIVISIONS = 100;
	private final static boolean DEFAULT_VISIBILITY = true;

	//Constructors
	protected Slider(Rect dimensions, Scalar color)
	{
		super(dimensions, color);
		init();
	}

	private void init()
	{
		numOfDivisions = DEFAULT_DIVISIONS;
		percentageVisible = DEFAULT_VISIBILITY;

		calculateDivisionSize();
	}

	//Setters
	protected void setNumOfDivisions(int numOfDivisions)
	{
		this.numOfDivisions = numOfDivisions;

		calculateDivisionSize();
	}

	protected void setPercentageVisible(boolean percentageVisible)
	{
		this.percentageVisible = percentageVisible;
	}

	//Getters
	protected int getNumOfDivisions()
	{
		return this.numOfDivisions;
	}

	protected boolean isPercentageVisible()
	{
		return this.percentageVisible;
	}

	private void calculateDivisionSize()
	{
		//Calculate the size of a single slider sector
		divisionSize = (double) dimensions.height / numOfDivisions;
	}

	@Override
	protected void updateDetectionPoint(Mat filteredImage)
	{
		super.updateDetectionPoint(filteredImage);

		if (hasDetection())
			division = (int) Math.ceil(((detectionPoint.y - dimensions.y) / divisionSize));
	}

	@Override
	protected void drawOnto(Mat image)
	{
		super.drawOnto(image);

		setSliderStatus(image);
	}

	private void setSliderStatus(Mat rawImage)
	{
		//Calculate position of line
		double linePosition = divisionSize * division + dimensions.y;

		//Draw status line
		Imgproc.line(rawImage, new Point(dimensions.x, linePosition),
				new Point(dimensions.x + dimensions.width, linePosition), color, 3);

		if (percentageVisible)
		{
			//Offset text to avoid collision with line and slider
			int textShift = (100 - division >= 50) ? 18 : -8;

			//Draw text
			Imgproc.putText(rawImage, 100 - division + "%", new Point(dimensions.x + 5, linePosition + textShift),
					Core.FONT_HERSHEY_COMPLEX, 0.5, color);
		}
	}

	@Override
	public String toString()
	{
		return super.toString() + String.format(TO_STRING_FORMAT, "Divisions:", this.numOfDivisions)
				+ String.format(TO_STRING_FORMAT, "Percent visible:", this.percentageVisible);
	}
}

File: src/main/kotlin/me/jakejmattson/touchcontrol/touchables/Slider.kt
package me.jakejmattson.touchcontrol.touchables

import org.opencv.core.*
import org.opencv.imgproc.Imgproc
import kotlin.math.ceil

/**
 * Abstract Slider - contains general slider behavior.
 *
 * @author JakeJMattson
 */
abstract class Slider protected constructor(dimensions: Rect, color: Scalar) : Touchable(dimensions, color) {
    /**
     * Number of sectors that the slider is divided into
     */
    protected var numOfDivisions: Int = DEFAULT_DIVISIONS
        set(value) {
            field = value.coerceIn(0..100)
        }

    /**
     * The size of a single division sector
     */
    private val divisionSize: Double
        get() = dimensions.height.toDouble() / numOfDivisions

    /**
     * The division in which the detection point is located in
     */
    protected var currentDivision: Int = 0

    /**
     * Whether or not the location line should have a label
     */
    private var isLabeled: Boolean = DEFAULT_VISIBILITY

    override fun updateDetectionPoint(filteredImage: Mat): Point? {
        super.updateDetectionPoint(filteredImage)

        if (hasDetection())
            currentDivision = ceil((detectionPoint!!.y - dimensions.y) / divisionSize).toInt()

        return detectionPoint
    }

    override fun drawOnto(image: Mat) {
        super.drawOnto(image)
        setSliderStatus(image)
    }

    private fun setSliderStatus(image: Mat) {
        //Calculate position of line
        val linePosition = divisionSize * currentDivision + dimensions.y

        //Draw status line
        Imgproc.line(image, Point(dimensions.x.toDouble(), linePosition),
            Point((dimensions.x + dimensions.width).toDouble(), linePosition), color, 3)

        if (isLabeled) {
            //Offset text to avoid collision with line and slider
            val textShift = if (numOfDivisions - currentDivision >= numOfDivisions / 2.0) 18 else -8

            //Calculate percentage to display
            val percent = (numOfDivisions - currentDivision) * (100 / numOfDivisions)

            val unit =
                when (numOfDivisions) {
                    100 -> "%"
                    else -> ""
                }

            //Draw text
            Imgproc.putText(image, percent.toString() + unit,
                Point((dimensions.x + 5).toDouble(), linePosition + textShift),
                Imgproc.FONT_HERSHEY_COMPLEX, 0.5, color)
        }
    }

    override fun toString() = super.toString() +
        format("Divisions:", numOfDivisions) +
        format("Is Labeled:", isLabeled)

    companion object {
        private const val DEFAULT_DIVISIONS = 100
        private const val DEFAULT_VISIBILITY = true
    }
}